WordPressKit/WordPressComRestApi.swift

```diff
@@ -299,18 +299,54 @@ open class WordPressComRestApi: NSObject {
         return "\(String(describing: oAuthToken)),\(String(describing: userAgent))".hashValue
     }
 
-    fileprivate func buildRequestURLFor(path: String) -> String? {
-        let pathWithLocale = appendLocaleIfNeeded(path)
+    // AlamoFire our parameters via `SessionManager.request(_:method:parameters:encoding:headers:)`
+
+    /// This method assembles a valid request URL for the specified path & parameters.
+    /// The framework relies on a field (`appendsPreferredLanguageLocale`) to influence whether or not locale should be
+    /// added to the path of requests. This approach did not consider request parameters.
+    ///
+    /// This method now considers both the path and specified request parameters when performing the substitution.
+    /// It only accounts for the locale parameter. AlamoFire encodes other parameters via `SessionManager.request(_:method:parameters:encoding:headers:)`
+    ///
+    /// - Parameters:
+    ///   - path: the path for the request, which might include `locale`
+    ///   - parameters: the request parameters, which could conceivably include `locale`
+    ///   - localeKey: the locale key to search for (`locale` in v1 endpoints, `_locale` for v2)
+    /// - Returns: a request URL if successful, `nil` otherwise.
+    ///
+    func buildRequestURLFor(path: String, parameters: [String: AnyObject]? = [:], localeKey: String = WordPressComRestApi.localeKey) -> String? {
+
         let baseURL = URL(string: WordPressComRestApi.apiBaseURLString)
-        let requestURLString = URL(string: pathWithLocale, relativeTo: baseURL)?.absoluteString
-        return requestURLString
+
+        guard let requestURLString = URL(string: path, relativeTo: baseURL)?.absoluteString,
+            let urlComponents = URLComponents(string: requestURLString) else {
+
+            return nil
+        }
+
+        let urlComponentsWithLocale = applyLocaleIfNeeded(urlComponents: urlComponents, parameters: parameters, localeKey: localeKey)
+
+        return urlComponentsWithLocale?.url?.absoluteString
     }
 
-    fileprivate func appendLocaleIfNeeded(_ path: String) -> String {
+    private func applyLocaleIfNeeded(urlComponents: URLComponents, parameters: [String: AnyObject]? = [:], localeKey: String) -> URLComponents? {
         guard appendsPreferredLanguageLocale else {
-            return path
+            return urlComponents
+        }
+
+        var componentsWithLocale = urlComponents
+        var existingQueryItems = componentsWithLocale.queryItems ?? []
+
+        let existingLocaleQueryItems = existingQueryItems.filter { $0.name == localeKey }
+        if let parameters = parameters, parameters[localeKey] == nil, existingLocaleQueryItems.isEmpty {
+            let preferredLanguageIdentifier = WordPressComLanguageDatabase().deviceLanguage.slug
+            let localeQueryItem = URLQueryItem(name: WordPressComRestApi.localeKey, value: preferredLanguageIdentifier)
```

Comment:
Should we use the passed in `localeKey` value here, not the constant directly?

If so, we could potentially add another test to cover this, passing a different `localeKey` to `buildRequestURLFor`.

Reply:
Good catch. I pushed a commit that sets the key to the value of the injected `localeKey`.

As for the test with different values, this is forthcoming via #78.